File: Extensions/Tessa.Extensions.Client/UI/PnrUKOutgoingUIExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Shared.PnrCards;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Client.UI
{
    public sealed class PnrUKOutgoingUIExtension : CardUIExtension
    {
        private readonly ICardRepository cardRepository;
        public PnrUKOutgoingUIExtension(ICardRepository cardRepository)
        {
            this.cardRepository = cardRepository;
        }

        private async void SetLegalEntityIndex(Guid organizationID, ICardModel cardModel)
        {
            CardRequest request = new CardRequest
            {
                RequestType = Shared.PnrRequestTypes.GetIndexLegalEntityRequestTypeID,
                Info =
                {
                    { "organizationID", organizationID }
                }
            };

            CardResponse response = await cardRepository.RequestAsync(request);

            Tessa.Platform.Validation.ValidationResult result = response.ValidationResult.Build();
            TessaDialog.ShowNotEmpty(result);
            if (result.IsSuccessful)
            {
                //cardModel.Card.Sections["PnrOutgoingUK"].Fields["LegalEntityIndexID"] = response.Info.Get<Guid?>("LegalEntityID");
                cardModel.Card.Sections["PnrOutgoingUK"].Fields["LegalEntityIndexIdx"] = response.Info.Get<string>("LegalEntityIdx");
            }
        }

        private async void SetUserDepartmentInfo(Guid? authorID, ICardModel cardModel)
        {
            CardRequest request = new CardRequest
            {
                RequestType = Shared.PnrRequestTypes.GetUserDepartmentInfoRequestTypeID,
                Info =
                {
                    { "authorID", authorID }
                }
            };

            CardResponse response = await cardRepository.RequestAsync(request);

            Tessa.Platform.Validation.ValidationResult result = response.ValidationResult.Build();
            TessaDialog.ShowNotEmpty(result);
            if (result.IsSuccessful && cardModel.Card.Sections["PnrOutgoingUK"].Fields["DepartmentID"] == null)
            {
                cardModel.Card.Sections["PnrOutgoingUK"].Fields["DepartmentID"] = response.Info.Get<Guid?>("DepartmentID");
                cardModel.Card.Sections["PnrOutgoingUK"].Fields["DepartmentName"] = response.Info.Get<string>("Name");
                cardModel.Card.Sections["PnrOutgoingUK"].Fields["DepartmentIdx"] = response.Info.Get<string>("Index");
            }
        }

        // установка видимости контролов, зависимых от Вида исходящего документа УК ПС
        private void SetDKDependenceVisibility(string documentKindIdx, ICardModel cardModel)
        {
            IBlockViewModel incomingDocsBlock = cardModel.Blocks["OutgoingDocsBlock"];
            incomingDocsBlock.BlockVisibility = documentKindIdx == PnrOutgoingUKTypes.OutgoingUKLetterIdx ? Visibility.Visible : Visibility.Collapsed;

            IBlockViewModel complaintsBlock = cardModel.Blocks["ComplaintsBlock"];
            complaintsBlock.BlockVisibility = documentKindIdx == PnrOutgoingUKTypes.OutgoingUKComplaintsIdx ? Visibility.Visible : Visibility.Collapsed;
            complaintsBlock.Rearrange();
        }

        public override Task Initialized(ICardUIExtensionContext context)
        {
            Card card = context.Card;
            ICardModel cardModel = context.Model;

            // Создание карточки
            if (card.StoreMode == CardStoreMode.Insert)
            {
                // Заполнить поле Дата регистрации текущей датой
                if (cardModel.Controls.TryGet("RegistrationDate", out IControlViewModel projectDateControl))
                {
                    ((DateTimeViewModel)projectDateControl).SelectedDate = DateTime.Now;
                }

                // начальная инициализация поля Подразделение по автору
                SetUserDepartmentInfo((Guid?)card.Sections["DocumentCommonInfo"].Fields["AuthorID"], cardModel);

                // поле Вид исходящего документа УК автозополнено - устанавливаем видимость контролов в зависимости от Вида
                SetDKDependenceVisibility((string)card.Sections["PnrOutgoingUK"].Fields["DocumentKindIdx"], cardModel);
            }
            else if (card.StoreMode == CardStoreMode.Update && card.Sections["PnrOutgoingUK"].Fields["DocumentKindID"] != null)
            // Открытие карточки. Есть Вид входящего документа - устанавливаем видимость контролов в зависимости от Вида
            {
                SetDKDependenceVisibility((string)card.Sections["PnrOutgoingUK"].Fields["DocumentKindIdx"], cardModel);
            }

            context.Card.Sections["PnrOutgoingUK"].FieldChanged += (s, e) =>
            {
                // Вид исходящего документа: Visibility контролов
                if (e.FieldName == "DocumentKindIdx" && e.FieldValue != null)
                {
                    SetDKDependenceVisibility((string)e.FieldValue, cardModel);
                }

                // Организация ГК Пионер: заполнение Индекс ЮЛ
                if (e.FieldName == "OrganizationID" && e.FieldValue != null)
                {
                    SetLegalEntityIndex((Guid)e.FieldValue, cardModel);
                }
            };

            return base.Initialized(context);
        }
    }
}

File: WebClient SDK/src/PnrScripts/UI/PnrUKOutgoingUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { Guid, createTypedField, DotNetType, Visibility } from 'tessa/platform';
import { CardStoreMode } from 'tessa/cards';
import { PnrCardTypes } from '../Shared/PnrCardTypes';
import { PnrRequestTypes } from '../Shared/PnrRequestTypes';
import { PnrOutgoingUKTypes } from '../Shared/PnrOutgoingUKTypes';
import { CardRequest, CardService } from 'tessa/cards/service';
import { DateTimeViewModel } from 'tessa/ui/cards/controls';
import Moment from 'moment';

export class PnrUKOutgoingUIExtension extends CardUIExtension {
  public async initialized(context: ICardUIExtensionContext) {
    if (!Guid.equals(context.card.typeId, PnrCardTypes.PnrOutgoingUKTypeID)) {
      return;
    }
    let PnrOutgoingUK = context.card.sections.tryGet("PnrOutgoingUK");
    if (!PnrOutgoingUK) return;
    // Создание карточки
    if (context.card.storeMode == CardStoreMode.Insert)
    {
      const regDateControl = context.model.controls.get('RegistrationDate') as DateTimeViewModel;
      if (!regDateControl) return;
      // Заполнить поле Дата регистрации текущей датой
      regDateControl.selectedDate = Moment(Date.now());

      let dci = context.card.sections.tryGet("DocumentCommonInfo");
      if (dci != null) {
        // начальная инициализация поля Подразделение по автору
        await this.SetUserDepartmentInfo(dci.fields.get('AuthorID'), context);
      }
      // поле Вид исходящего документа автозополнено - устанавливаем видимость контролов в зависимости от Вида
      this.SetDKDependenceVisibility(PnrOutgoingUK.fields.get('DocumentKindIdx'), context);
    }
    else if (PnrOutgoingUK.fields.get('DocumentKindID') != null) {
      // Открытие карточки. Есть Вид исходящего документа - устанавливаем видимость контролов в зависимости от Вида
      this.SetDKDependenceVisibility(PnrOutgoingUK.fields.get('DocumentKindIdx'), context);
    }

    PnrOutgoingUK.fields.fieldChanged.add(async (e) => {
      // Вид исходящего документа: Visibility контролов и переформирование номера
      if (e.fieldName == "DocumentKindIdx" && e.fieldValue != null)
      {
         this.SetDKDependenceVisibility(e.fieldValue, context);
      }
      // Организация ГК Пионер: заполнение Индекс ЮЛ + изменение Номера
      if (e.fieldName == "OrganizationID" && e.fieldValue != null)
      {
        await this.SetLegalEntityIndex(e.fieldValue, context);
      }
    });
  }

  private async SetLegalEntityIndex(organizationID, context: ICardUIExtensionContext)
  {
    // request к организации получить индекс ЮЛ
    let LegalEntityIdx;
    let request = new CardRequest();
        request.requestType = PnrRequestTypes.GetIndexLegalEntityRequestTypeID;
        request.info["organizationID"] = createTypedField(organizationID, DotNetType.Guid);
    // проверка ответа
    let response = await CardService.instance.request(request);
    if (response.validationResult.isSuccessful) {
        LegalEntityIdx = response.info.LegalEntityIdx;
    }
    let PnrOutgoingUK = context.card.sections.tryGet("PnrOutgoingUK");
    if (LegalEntityIdx != null && PnrOutgoingUK != null) {
      PnrOutgoingUK.fields.set('LegalEntityIndexIdx', createTypedField(LegalEntityIdx.$value, DotNetType.String))
    }
  }

  private async SetUserDepartmentInfo(authorID, context: ICardUIExtensionContext)
  {
    let DepartmentID;
    let DepartmentName;
    let DepartmentIdx;
    let request = new CardRequest();
        request.requestType = PnrRequestTypes.GetUserDepartmentInfoRequestTypeID;
        request.info["authorID"] = createTypedField(authorID, DotNetType.Guid);

    // проверка ответа
    let response = await CardService.instance.request(request);
    if (response.validationResult.isSuccessful) {
      DepartmentID = response.info.DepartmentID;
      DepartmentName = response.info.Name;
      DepartmentIdx = response.info.Index;
    }

    let PnrOutgoingUK = context.card.sections.tryGet("PnrOutgoingUK");
    if (DepartmentID != null && DepartmentName != null && DepartmentIdx != null && PnrOutgoingUK != null && PnrOutgoingUK.fields.get('DepartmentID') == null) {
      PnrOutgoingUK.fields.set('DepartmentID', createTypedField(DepartmentID.$value, DotNetType.Guid));
      PnrOutgoingUK.fields.set('DepartmentName', createTypedField(DepartmentName.$value, DotNetType.String));
      PnrOutgoingUK.fields.set('DepartmentIdx', createTypedField(DepartmentIdx.$value, DotNetType.String));
    }
  }

  /** Установка видимости контролов, зависимых от Вида исходящего документа УК ПС */
  private SetDKDependenceVisibility(documentKindIdx, context: ICardUIExtensionContext)
  {
    const incomingDocsBlock = context.model.blocks.get('OutgoingDocsBlock');
    if (incomingDocsBlock) {
      incomingDocsBlock.blockVisibility = documentKindIdx == PnrOutgoingUKTypes.OutgoingUKLetterIdx ? Visibility.Visible : Visibility.Collapsed;
    }
    const complaintsBlock = context.model.blocks.get('ComplaintsBlock');
    if (complaintsBlock) {
      complaintsBlock.blockVisibility = documentKindIdx == PnrOutgoingUKTypes.OutgoingUKComplaintsIdx ? Visibility.Visible : Visibility.Collapsed;
    }
  }
}